File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/persistence/JpaCategoryRepository.java
package com.zappymart.adapters.persistence;

import com.zappymart.application.ports.CategoryRepository;
import com.zappymart.domain.catalogue.Category;

import java.util.List;

public final class JpaCategoryRepository implements CategoryRepository {

    private final CategoryRowRepository categoryRows;

    JpaCategoryRepository(CategoryRowRepository categoryRows) {
        this.categoryRows = categoryRows;
    }

    @Override
    public List<Category> inCatalogueOrder() {
        return categoryRows.findAllByOrderByPositionAsc().stream()
                .map(row -> new Category(row.id, row.name, row.slug))
                .toList();
    }
}

File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/persistence/JpaCategoryRepository.kt
package nl.zappymart.adapters.persistence

import nl.zappymart.application.ports.CategoryRepository
import nl.zappymart.domain.catalogue.Category
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
@Transactional(readOnly = true)
class JpaCategoryRepository(private val categories: CategoryEntities) : CategoryRepository {

    override fun findAll(): List<Category> =
        categories.findAllByOrderByPositionAsc().map { entity -> entity.asCategory() }
}
